### PR TITLE
[ONNX] Support per channel quantization

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76055
* __->__ #76002

Extending the support for quantization with per channel quantization.
An extra attribute `axis` can be found for per channel quantized tensors,
most commonly in quantized weight of Convolution or Linear module.
The PR adds support to correctly parse the `axis` attribute, and map to
ONNX representation in `QuantizeLinear` and `DequantizeLinear`.

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -10741,6 +10741,141 @@ def forward(self, x):
...
+        model = M()
+        model.qconfig = torch.quantization.get_default_qconfig("fbgemm")
+        model = torch.quantization.prepare_qat(model)
+        # Set fixed weight and bias to avoid flaky test.
+        model.conv.weight = torch.nn.Parameter(torch.arange(4 * 2 * 3 * 3, dtype=torch.float).view(2, 4, 3, 3) / 36)
+        model.conv.bias = torch.nn.Parameter(torch.arange(2, dtype=torch.float))
+        model = torch.quantization.convert(model)
+
+        # Set fixed input to avoid flaky test.
+        input = (torch.arange(3 * 4 * 8 * 8, dtype=torch.float).view(3, 4, 8, 8) - 384) / 32
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:10817]: `(torch.arange(x).view() - y) / z` seems to be a repeated pattern, and it's not clear what the relationships are between the args to `arange`, `view`, `-` and `/`.
Could you write a little helper that names the variables and explains it, and then call the helper function in all these cases?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -10741,6 +10741,141 @@ def forward(self, x):
...
+            def forward(self, x):
+                x = self.quant(x)
+                x = self.conv(x)
+                x = self.relu(x)
+                x = self.dequant(x)
+                return x
+
+        model = M()
+        model.qconfig = torch.quantization.get_default_qconfig("fbgemm")
+        model = torch.quantization.fuse_modules(model.eval(), [['conv', 'relu']])
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:10868]:
```suggestion
        model = torch.quantization.fuse_modules(model.eval(), [["conv", "relu"]])
```

File: torch/onnx/symbolic_helper.py [hunk 3]
```diff
@@ -950,29 +963,58 @@ def dequantize_helper(g, qtensor, qdtype=None):
...
+    output = g.op("QuantizeLinear", tensor, scale, zero_point, **axis_kwargs)
+    args = [output, scale, zero_point]
+    if axis is not None and not _is_none(axis):
+        args.append(axis)
+    return g.op("prim::TupleConstruct", *args)
+
+def requantize_bias_helper(g, bias, input_scale, weight_scale, axis=None):
+    """In PyTorch, bias is float and is quantized implicitly inside the quantized ATen op kernel.
+    In ONNX we need to make the quantization explicit because operators expect all of their inputs to be quantized.
+    Since int32 is not supported by ONNX operator `QuantizeLinear`, quantization is exported using regular operators.
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:1006]: Which input or output of `QuantizeLinear` would we need to support int32?
Please expand the comment to be more specific.

File: torch/onnx/symbolic_helper.py [hunk 4]
```diff
@@ -950,29 +963,58 @@ def dequantize_helper(g, qtensor, qdtype=None):
...
+    if _export_onnx_opset_version < 13:
+        if axis_i is not None:
+            _onnx_opset_unsupported_detailed(
+                "DequantizeLinear", _export_onnx_opset_version, 13, "Attribute axis is not supported.")
+
+    axis_kwargs = {"axis_i": axis_i} if axis_i is not None else {}
+    return g.op("DequantizeLinear", value, scale, zero_point, **axis_kwargs), scale, zero_point, axis
+
+def quantize_helper(g, tensor, scale, zero_point, axis=None):
+    """Append to graph `g` with ONNX nodes that quantizes `tensor` based on `scale`, `zero_point` and `axis`.
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:976]:
```suggestion
    """Append to graph `g` ONNX nodes that quantizes `tensor` based on `scale`, `zero_point` and `axis`.
```

File: torch/onnx/symbolic_helper.py [hunk 5]
```diff
@@ -940,7 +940,20 @@ def _handle_reduce_dim_none(g, self, op_name):
     return g.op(op_name, self, keepdims_i=0)
 
 def dequantize_helper(g, qtensor, qdtype=None):
-    tensor, scale, zero_point = _unpack_tuple(qtensor)
+    """Append to graph `g` with ONNX nodes that dequantizes `qtensor` into `tensor`.
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:943]:
```suggestion
    """Append to graph `g` ONNX nodes that dequantizes `qtensor` into `tensor`.
```

File: torch/onnx/symbolic_helper.py [hunk 6]
```diff
@@ -950,29 +963,58 @@ def dequantize_helper(g, qtensor, qdtype=None):
...
     if zero_point.type().scalarType() not in ("Byte", "Char"):
         zero_point = g.op("Cast", zero_point, to_i=torch.onnx.TensorProtoDataType.UINT8)
-    output = g.op("QuantizeLinear", tensor, scale, zero_point)
-    return g.op("prim::TupleConstruct", output, scale, zero_point)
-
-def requantize_bias_helper(g, bias, input_scale, weight_scale):
-    # In PyTorch, bias is float and is quantized implicitly inside the quantized ATen op kernel.
-    # In ONNX we need to make the quantization explicit because operators expect all of their inputs to be quantized.
-    # Since int32 is not supported by ONNX operator `QuantizeLinear`, quantization is exported using regular operators.
+    output = g.op("QuantizeLinear", tensor, scale, zero_point, **axis_kwargs)
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:997]: It seems like `axis_kwargs` is set unconditionally. In which case you can get rid of it and just always specify `axis_i`. Or is it a bug that `axis_kwargs` is always set?

```suggestion
    output = g.op("QuantizeLinear", tensor, scale, zero_point, axis_i=_get_const(axis, "i", "axis"))
```

File: torch/onnx/symbolic_helper.py [hunk 7]
```diff
@@ -950,29 +963,58 @@ def dequantize_helper(g, qtensor, qdtype=None):
     value = g.op("Cast", tensor, to_i=qdtype)
     scale = g.op("Cast", scale, to_i=torch.onnx.TensorProtoDataType.FLOAT)
     zero_point = g.op("Cast", zero_point, to_i=qdtype)
-    return g.op("DequantizeLinear", value, scale, zero_point), scale, zero_point
 
-def quantize_helper(g, tensor, scale, zero_point):
+    if _export_onnx_opset_version < 13:
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:967]: Suggestion for making the control flow a bit clearer here. 

```python
axis_kwargs = {}
if axis_i:
  if _export_onnx_opset_version < 13:
     _onnx_opset_unsupported_detailed(...)
   axis_kwargs = {"axis_i": axis_i}
```

File: torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp [hunk 8]
```diff
@@ -175,6 +175,100 @@ std::vector<Node*> CreateQuantizedWeights(
   return {const_node_1, const_node_2, const_node_3};
 }
 
+std::vector<Node*> CreateQuantizedWeights(
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:178]: Can you add comments to this and the other function that has the same name on line 132 explaining the differences between the two? Or maybe name them differently.

File: torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp [hunk 9]
```diff
@@ -175,6 +175,100 @@ std::vector<Node*> CreateQuantizedWeights(
...
+    std::shared_ptr<Graph>& graph,
+    const at::Tensor& weight,
+    int8_t* data,
+    const std::vector<int64_t>& shapes,
+    const std::vector<int64_t>& strides) {
+  auto qschema = weight.qscheme();
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:184]: schema -> scheme for consistency with with PyTorch nomenclature.
Also update all relevant comments and strings below.

File: torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp [hunk 10]
```diff
@@ -175,6 +175,100 @@ std::vector<Node*> CreateQuantizedWeights(
...
+    case c10::kPerChannelAffineFloatQParams: {
+      auto q_scales = weight.q_per_channel_scales();
+      auto* scale_data_raw =
+          reinterpret_cast<double*>(q_scales.data_ptr<double>());
+      scale_shapes = q_scales.sizes().vec();
+      TORCH_INTERNAL_ASSERT(
+          scale_shapes.size() == 1,
+          "quantized per channel scales are expected as 1-d array.");
+      scale_data.resize(scale_shapes[0]);
+      std::transform(
```
Comment (garymm) [inline torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:212]: can you add a comment explaining why we need floats rather than doubles?

**Review by garymm — APPROVED** [review]

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 11]
```diff
@@ -10741,6 +10741,141 @@ def forward(self, x):
         x = torch.quantize_per_tensor(torch.randn(3, 4), 0.2, 0, torch.qint8)
         self.run_test(Module(), x)
 
+    @skipIfUnsupportedMinOpsetVersion(13)
+    def test_qat_linear_per_channel(self):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:10745]: it is `torch.quantization.get_default_qconfig("fbgemm")`, which leads to

```python
        if backend == 'fbgemm':
            qconfig = QConfig(activation=HistogramObserver.with_args(reduce_range=True),
                              weight=default_per_channel_weight_observer)
```

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 1 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:10817]: It's really just trying to deterministically construct a tensor with different element values within reasonable range. Perhaps a handwritten constant tensor is better?

File: torch/onnx/symbolic_helper.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py:1006]: Changed the first line of comment to "... is quantized **to int32** implicitly ...", and the last line to "Since int32 is not **a supported output type** by ..."

It is bias in this case.

File: torch/onnx/symbolic_helper.py (hunk 6 above)
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py:997]: Good catch. It turns out for passing attribute, the attribute will not be set if value being passed is None. Hence we can get rid of the kwargs and always specify `axis_i`.

File: torch/onnx/symbolic_helper.py (hunk 7 above)
Comment (BowenBao) [reply torch/onnx/symbolic_helper.py:967]: Related to the other comment, can simplify condition to one line now that we don't need axis_kwargs.

File: torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp (hunk 8 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:178]: forgot to remove old code ...

File: torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp (hunk 9 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:184]: done

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 12]
```diff
@@ -10779,6 +10779,141 @@ def forward(self, x):
...
+        class M(torch.nn.Module):
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:10784]: Is this module the same across all new test cases? If so would it be a good idea to extract the class to the global namespace?

```
class _TestQuatizedModule(torch.nn.Module):
    ...
```

File: torch/onnx/symbolic_helper.py [hunk 13]
```diff
@@ -957,7 +957,20 @@ def _handle_reduce_dim_none(g, self, op_name):
     return g.op(op_name, self, keepdims_i=0)
 
 def dequantize_helper(g, qtensor, qdtype=None):
-    tensor, scale, zero_point = _unpack_tuple(qtensor)
+    """Append to graph `g` ONNX nodes that dequantizes `qtensor` into `tensor`.
```
Comment (justinchuby) [inline torch/onnx/symbolic_helper.py:960]: nit: consistency: Append**s**

File: torch/onnx/symbolic_helper.py [hunk 14]
```diff
@@ -967,29 +980,56 @@ def dequantize_helper(g, qtensor, qdtype=None):
...
 
-def quantize_helper(g, tensor, scale, zero_point):
+    if axis_i is not None and _export_onnx_opset_version < 13:
+        _onnx_opset_unsupported_detailed(
+            "DequantizeLinear", _export_onnx_opset_version, 13, "Attribute axis is not supported.")
+
+    return g.op("DequantizeLinear", value, scale, zero_point, axis_i=axis_i), scale, zero_point, axis
+
+def quantize_helper(g, tensor, scale, zero_point, axis=None):
+    """Append to graph `g` ONNX nodes that quantizes `tensor` based on `scale`, `zero_point` and `axis`.
```
Comment (justinchuby) [inline torch/onnx/symbolic_helper.py:991]: nit: Appends

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 15]
```diff
@@ -10847,12 +10866,12 @@ def forward(self, x):
         model.qconfig = torch.quantization.get_default_qconfig("fbgemm")
         model = torch.quantization.prepare_qat(model)
         # Set fixed weight and bias to avoid flaky test.
-        model.conv.weight = torch.nn.Parameter(torch.arange(4 * 2 * 3 * 3, dtype=torch.float).view(2, 4, 3, 3) / 36)
-        model.conv.bias = torch.nn.Parameter(torch.arange(2, dtype=torch.float))
+        model.conv.weight = torch.nn.Parameter(_construct_tensor_for_quantization_test((2, 4, 3, 3), max_val=2))
+        model.conv.bias = torch.nn.Parameter(torch.tensor([0., 1.]))
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:10870]: nit: Prefer adding the `0` for making a number float. `0.0` makes it apparent to readers that the number is a float.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 16]
```diff
@@ -283,6 +283,25 @@ def _init_test_roi_heads_faster_rcnn():
         box_score_thresh, box_nms_thresh, box_detections_per_img)
     return roi_heads
 
+def _construct_tensor_for_quantization_test(shape, offset=None, max_val=None):
+    """Due to difference in implementation details between PyTorch and ONNXRuntime, randomly generated
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:287]: Very much a nit, but: Prefer starting the doc with a single line summary: https://google.github.io/styleguide/pyguide.html#381-docstrings

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 17]
```diff
@@ -283,6 +283,25 @@ def _init_test_roi_heads_faster_rcnn():
...
+    test data for quantization tests can be flaky. To help stablize the test, this helper function is
+    used to generate weights and test inputs in a deterministic way.
+
+    Args:
+        shape (Tuple[int]): Shape for tensor to construct.
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:292]: Adding the type annotations inline will help IDEs analyze the code. Type checkers like pylance will use this info to display errors inline if there is a typing issue. This will ultimately help us understand the code better and code for efficiently. Happy to discuss further.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 15 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:10870]: is there a case where `0.` could be not floating number? I feel the current style is equally readable while being less verbose.

Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:10875]: 0. is perfectly fine syntactically. I am fine with both ways and in fact wrote it like so before :). I received this suggestion in a past readability review so I was passing it on.

Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:10875]: Discussions are always welcomed :). For this particular case I think both ways are fine. The `0.` approach makes tensors/arrays more compact which is in my opinion better for displaying larger tensors in code, hence followed as convention.

Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:10875]: SGTM!

File: torch/onnx/symbolic_opset13.py [hunk 18]
```diff
@@ -417,3 +418,43 @@ def diagonal(g, self, offset, dim1, dim2):
...
     _add_output_to_block(else_block, final_overrun_)
     return if_op
+
+class Quantized:
+    """
+    https://github.com/pytorch/pytorch/wiki/PyTorch-ONNX-exporter#quantized-model-export
+    """
+    domain = "quantized"
+
+    @staticmethod
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset13.py:428]: non-blocking: Is there a way we don't need to staticmethods, or are there special considerations?

https://google.github.io/styleguide/pyguide.html#217-function-and-method-decorators and https://mail.python.org/pipermail/python-ideas/2016-July/041189.html from Guido himself.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 19]
```diff
@@ -283,6 +283,30 @@ def _init_test_roi_heads_faster_rcnn():
         box_score_thresh, box_nms_thresh, box_detections_per_img)
     return roi_heads
 
+def _construct_tensor_for_quantization_test(shape: Tuple[int],
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:286]: shape: Tuple[int, ...]

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 15 above)
Comment (garymm) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:10875]: +1 for `0.`.

File: torch/onnx/symbolic_opset13.py (hunk 18 above)
Comment (garymm) [reply torch/onnx/symbolic_opset13.py:428]: We could switch to using a module rather than a class for each domain.
But neither of those links explains why we should not use static methods.
Feel free to send a PR to change to modules if you can provide a good justification.

Comment (justinchuby) [reply torch/onnx/symbolic_opset13.py:428]: The main argument I know of is that we usually write static methods in a class just for a name space, whereas the Pythonic way of name spacing is via modules. I assume things can be different if we wanted to implement some kind of protocols with the class, overwriting the same signatures for example. In that case the class will be a reference when used and its name not spelled out in the caller. I wonder if that would add obscurity to the code and if that is something we were trying to do with this class.
